### ISSUE TITLE
Initialization / run / postprocess

### ISSUE THREAD

**Issue by denisri** [populse/populse_mia#181]
Within the context of clarifying what MIA should do to run pipelines, and linked to #167 and #180, I open a different issue not to add confusion on existing issues, and more to discuss what should be done.

We already have an initialization function, and a postprocess function (in #167 at least). I have questions about what and when things should be done:

initialization
---------------
* check that the pipeline is OK
* check requirements and config
* complete parameters
* index processes/bricks and future output data ? This is done by this time in the current code. Processes which will run are registered in the database, and all their ouput data also. A record of their init time is kept.
  However this is not clear in my mind why this has to be done before running. Processes have not run yet; output data do not exist yet, so why registering them in the database ? Why is it interesting/important to record init time ?
  A pipeline may be initialized several times (with or without being run later), and currently we get an growing list of init records for the same (possibly non-existing) data. Is it really OK ? I feel it's confusing for the user, and things get messy...

postprocess
----------------
* Processing nodes are postprocessed: if they have a `manage_brick_after_run` method, then it is called)
* data indexation: currently this is done during init. But I feel output data should be indexed after the processing has ended, and when data actually exist, so this may be the right time ?
* processes indexation: as for data currently this is done during init, but I feel this should be done after the process has run.

which pipeline(s) to postprocess
-------------------------------------------
For init, this is rather clear: init a pipeline, then run it (although there is the multiple init issue mentioned above).
For postprocessing it is less clear. In a "synchronous" run, the client waits for pipeline execution to end, then postprocesses it - this is the "simple" case. But in a remote / asynchronous execution, the execution is handled in a soma-workflow server, and the client (mia) can be closed, the processing goes on and eventually finishes, then the client may be reconnected later. Several runs may be launched. So the postprocessing procedure should not be limited to a single, "current" pipeline run, but recover all runs that have been started.
Moreover a pipeline run may have been stopped (by user interruption, technical problem) or failed, and restarted after a first postprocessing. In my opinion a pipeline run is still living as long as its workflow is still in soma-workflow database (or at least it completes entierly without error).
How do we know/decide which pipelines have to be postprocessed ? How do we remember them after a client quit/restart ? Do we index runs in MIA database ? Or in soma-workflow ?

**Comment by sapetnioc** [comment]
I think that the database record corresponding to a process should be evolving during its life time. To date this record is created at initialization time, I think this is interesting only for successful initialization. It allows to query all the processes that are ready to start. If a user initialize twice the same process, Mia can see it with the database and eventually ask the user for creating a new entry or replacing the existing one. When the process is started, its database status should be put to running and soma-workflow related information added to the process record.

We can define some moments where:
- database is checked for initialized processes that were never started
- check that running processes are finished by querying soma-workflow. Set an appropriate error state is soma-workflow cannot be contacted or the process has disappeared.
- Insert output values in database for all terminated processes not yet postprocessed
During this procedure, questions could be asked to the user when database cleanup is necessary. This could be done, for instance, done in the background at Mia startup.

I think we should not use soma-workflow to index processes because it can be unavailable and we need to store higher level information (I believe populse_db is more adapted).

**Comment by denisri** [comment]
Currently during postprocessing, there is no obvious way to determine which processes have actually run, if it went on correctly or not, or if an output file has really been produced by the right process execution.
Processing "bricks" get an UUID during init, it is used to record bricks in MIA database. But no link to actual Process instances to execution is maintained.
In order to keep track of everything, we should:
- keep the UUID in each process instance. This ID would be valid between init and postprocessing, and should be reset after a postprocessing stage (or before a new init), to get a different one if another run (init/exec) is done for the same pipeline instance. IDs would be both in database entries and Process instances so we can make the link.
- the UUIDs should also go into Soma-Workflow jobs, in order to keep links between real executions and MIA database
- During postprocessing, we can get the UUID of jobs in Soma-Workflow, recover the full execution logs/status (we can keep it in MIA databasse if we want to keep all the details - but it may be much data as some processes are very verbose), and update output files status according to jobs execution status. We can also get the real execution date from it.

It should be noted that:
- Process instances may be lost in client/server mode since the client can be closed and pipeline instances are not persistent. So postprocessing should not rely heavily on them.
- Soma-workflow workflows may be deleted by hand by the user, so we are not completely sure to recover them after a run during the postprocessing step. In such a situation we cannot update files status. So it should not be harmful.
- Database entries are created in MIA during the init step. If several init are done on the same pipeline instance, previous ones are not erased. We should probably clean things up - but how do we know what can be cleaned ?
- When closing MIA, database entries are removed for processes / files which have not actually run (process status is "Not Done"). This cleanup is helpful in most cases, but if we want (and we will need) to allow client disconnection while a pipeline is running remotely, this cleanup should not be done at that moment. When should it happen ?

**Comment by denisri** [comment]
Update on the current state on the branch populse_db2_capsulengine:
- UUIDs are now set in processes and workflow jobs
- in the postprocessing steps, soma-workflow jobs and current pipeline processes are all collected to update their status
- data history is updated, and obsolete runs are deleted from the database (if a data has been re-written by a more recent process run, the earlier history is useless)

However there are still a few flaws:
- data history cannot be entirely reconstructed without possibility of error: a data history may be complex and multiple (the data may have been overwritten several times, and in some cases modified, thus part of its older history should still be kept and valid). As we have no notion of processing chains in the history (and in fact the user may have run several processes / pipelines independently modifying the same data). The only criterion we can use is processing date, which is approximative (machines clocks may differ, and we cannot know when the data was actually written during the execution of jobs, when several of the write the same data in approximately the same time)
- execution date/time becomes an important criterion in the history, and we only have it when soma-workflow workflows have not been deleted. Thus in some cases we may lose this information.
- we cannot currently cleanup all obsolete runs, because some (SPM) have written script files, which are still in the database and reference a process run. We have to handle them specifically.

**Comment by denisri** [comment]
I think I quite agree with @servoz: I see no strong reason to index in the database things or events which have not happened yet. As we discussed they introduce much possible confusion and complications, and who will use information about an init for a job that has not run, or a data which has not been written (yet or never). Thus I would tend to insert things in the database only when they are done: when a data has been written, and a process would be indexed only when it has actually produced an output.
As said before this will require a bit of work to change this, but this will be mainly for simplification.

I have not looked at the code for several months and I don't have everything in mind now, and the only question I'm not sure is the retreival of processes "instances" at postprocessing time: if the client (GUI) was closed during processing, then the only information we can get after reconnection is from soma-workflow. In SWF we have jobs (which do not necessarily completely match a process execution, but nearly), but jobs don't contain the original process and parameters they were built from. We need a means to make such process instances "temporarily persistent" (during run, between the instant we start running and the postprocessing step time) until we index them "definitely" in the history if they have run successfully.

Up to now processes are indexed at init time, which make them persistant for this period, but they are "too persistant" then because they may not run, or fail during run.
* One possibility is to index them the same way as they are now during init, but when we actually start processing. This would avoid multiple indexation due to multiple initialization, but we would still store in the database "things from the future": processes that will perhaps not run, or fail, because the workflow execution may not run completely for any kind of reason.
* Or do we store them in a "client data" in soma-workflow (something that SWF will store without interpreting it) ? It would solve the "things from the future" issue, but if we lose a workflow (if it gets deleted on the server) we cannot index things at all.

**Comment by servoz** [comment]
I have no idea what to do if a workflow is deleted in the server.
This case seems to me difficult to manage if we index after the process, which would be the cleanest way (no more mess, because in this case we index only what is launched).

Maybe a solution for this extreme case is to continue as now with this only difference:
There is always a backup of the database before any initialisation, say the db0 object. 
We do a first initialisation, we arrive at db1, that's what is displayed in the databrowser.
We make a new initialisation without having made a run: we start again from db0, we arrive at db1, etc ... Until we make a run. When we launch a run db1 becomes db0.

In fact it's a very similar result to doing this, as in your first point, just before the processing starts. The only difference is that when we run just before the processing, we don't see the result of the initialisation (I confess that I have the habit of going to see the result of the initialisation in the databrowser before launching the run but I'm not sure that a visual examination is the best thing in the case of a large number of data !!!)  ....

**Comment by servoz** [comment]
I just did some tests and the sequential calculation doesn't really seem to work now.
I'm not surprised because, in the early days, we developed mia for local and sequential use!
I confess that until now, given the size of the work to start this project, against the available forces,  I rather concentrated on local and sequential use.
Now that the first sufficiently solid bases of mia exist, I think that it is indeed time to start effort in the direction of an also asynchronous and parallel use.
But in this case we will add items to the todo list which is already well loaded.
What I wanted to avoid until now is to have a nice potential project but never really functional.
That's why I focused on a local and sequential use.
Modulo, some points (some of them are not minor) to improve with V2 (I still have to finish opening several ticket about this) we can consider that the goal of a functional mia in local and sequential is now reached.
So let's go for the remote and parallel features ... :-)

**Comment by servoz** [comment]
Ah !!! and soma-worfkow doesn't seem to want to wait until the input data of a pipeline that doesn't exist yet, (my question above ; case of 2 tabs, so two independent pipelines, and the second one takes the result of the first).
This seems normal to avoid that soma-workflow waits indefinitely for the missing input data.
Or soma-workflow would be aware of the output of the 2 pipelines (with the order of each), which finally results in having only one pipeline... ?

**Comment by denisri** [comment]
> Can Soma-workflow handle a situation where input data does not yet exist? When it is in the same pipeline, it's ok, but if it is in the input of the pipeline, somaworkflow waits until the data exists?

No, Soma-Workflow does not wait for data, except in the specific case of input files transfers between the client filesystem and the computing server filesystem. It is not designed for this kind of use (at least not currently).
Soma-Workflow deals with dependencies between jobs, not between files, and does not really take care of the existence of files. If a job starts without its input files, it will fail and exit with an error.

> Ah !!! and soma-worfkow doesn't seem to want to wait until the input data of a pipeline that doesn't exist yet, (my question above ; case of 2 tabs, so two independent pipelines, and the second one takes the result of the first).
This seems normal to avoid that soma-workflow waits indefinitely for the missing input data.
Or soma-workflow would be aware of the output of the 2 pipelines (with the order of each), which finally results in having only one pipeline... ?

No right now, a workflow has to be "self contained" and independent. If there is a dependency between two pipelines, then it should be one bigger pipeline.
Of course we can add new features in soma-workflow, the simplest being to have "wait for file" kinds of jobs, but they would need to handle timeouts, or even failures in the previous pipeline, which is not really the spirit of the library. Another way would be to dynamically add jobs in a running pipeline. This would allow the dependencies between jobs to be checked.

**Comment by denisri** [comment]
> Can Soma-workflow handle a situation where input data does not yet exist?

Normally this is the job of pipeline initialization to check that input files exist. In the use case you describe here this requirement is not met and you just cannot initialize the second pipeline until the first one has produced the required outputs.

**Comment by denisri** [comment]
> I just did some tests and the sequential calculation doesn't really seem to work now.
> I'm not surprised because, in the early days, we developed mia for local and sequential use!

Currently (in v2) I have unplugged the input data selection for iterations (iterate over / select tag buttons). The "iterate pipeline" changes the pipeline to another pipeline with an iteration node, in order to have one pipeline with iterations inside it. Thus it still needs work for data selection. Apart from that, once data are selected and the iteration initialized, it runs in soma-workflow with parallel calculations when they can be done (and typically independent iteration steps can run in parallel). I've just checked that this is working.

> I confess that until now, given the size of the work to start this project, against the available forces, I rather concentrated on local and sequential use.
> Now that the first sufficiently solid bases of mia exist, I think that it is indeed time to start effort in the direction of an also asynchronous and parallel use.
> But in this case we will add items to the todo list which is already well loaded.
> What I wanted to avoid until now is to have a nice potential project but never really functional.
> That's why I focused on a local and sequential use.

I understand this point and agree: if the GUI at first is limited to sequential execution, it's OK for me as long as it doesn't forbid to extend this later. The Capsul/Soma-workflow infrastucture handle parallel execution, so we just have to take care of not breaking this capability. The database rollback, typically, is something (in my opinion) that we should not use for this reason, because it will make it very difficult to reintroduce parallelism later.

> Modulo, some points (some of them are not minor) to improve with V2 (I still have to finish opening several ticket about this) we can consider that the goal of a functional mia in local and sequential is now reached.
> So let's go for the remote and parallel features ... :-)

For me the "only" big feature to fix/reimplement in v2 is data selection for iterations. The rest is already working in parallel execution (inside a single workflow). Using several dependent tabs will be another thing, that we can leave for v3 (or later).

**Comment by servoz** [comment]
I agree, and my intention is not to add more features to soma-workflow but rather to check that what we imagine in mia is finally practicable with soma-workflow.
For example it is useless to allow mia to prepare (setup + initialisation) two pipelines in 2 different tabs at the same time, where the second one takes the output of the first one as input (in fact it is like the same pipeline, conceptually, so we have to design it  in the same tab !), and then launch them in parallel, if this is not possible in soma-workflow !
 but I agree with what you said and the idea of making database rollback is not the right one :-(((((

**Comment by denisri** [comment]
> For example it is useless to allow mia to prepare (setup + initialisation) two pipelines in 2 different tabs at the same time, where the second one takes the output of the first one as input (in fact it is like the same pipeline, conceptually, so we have to design it in the same tab !), and then launch them in parallel, if this is not possible in soma-workflow !

We could think about it later, why not, but let's say it's not a priority for now.

**Comment by servoz** [comment]
Uh ... I never thought about it before you mentioned it...

**Comment by servoz** [comment]
After thinking about it  ...

- we may want to set-up pipelines in 2 different tabs and initialise them without running them. This excludes the idea of database revert between two initialisations, if pipelines was no run.

- So, the only way to avoid having mess in the database in the case of multiple initialisations without a run is, as @denisri  noted,  to do the indexing at the time of the run (just before or just after).

- On the other hand, I think that we should keep all over current operations during initialisation (controller completion, etc ...)  so that the user can have a look at the parameters if they have been automatically changed. This corresponds well to the idea of initialisation which allows to check that everything is well prepared and that the run has all the chances to work well.

- What remains is the indexing (of outputs and processes). 

- The cleanest solution would be to do it just after the process has run. The problem is that if the workflow is lost in the case of remote use, the indexing is lost, even if the calculation has run well.

- The other solution proposed by @denisri  is to do it just before the run. In this case the risk is the opposite. We index data coming from the future, and if the calculation fails we will have a database containing data that does not exist.

-  However, we want the database to be as synchronized as possible with the existing data.

- Unless a new idea allows us to get around the problem if we index after the run (that seems the cleanest solution), there is no solution that allows a good result every time (after or before the run).

- It seems to me that the least worst solution is to do the indexing just before the run. The Exec tag makes sense here (we won't need the Init and Init Time tag anymore). We know, in this case, that we have output data indexed in database coming from the future (which does not exist yet and which will never exist in case of problem). So we define an "Exec time" tag which corresponds to the indexation, i.e. just before the launch of the brick and an "Exec Status" tag (it's the idea of the current Exec but more explicit) which is indexed to "running" (for example). Once the brick has finished running, there is a post-processing indexation which sets Exec Status to "Done". So, if the workflow is lost on the server Exec Status will remain at Running. We can even imagine a cutup time where Exec Status is set to Fall. The observation of the two tags could indicate that there is a problem (if after 10 years the status is still on running, there is a doubt ...). 
In my opinion, this is an option to manage a situation where there doesn't seem to be a solution that works 100% for the moment.

**Comment by denisri** [comment]
>     * The cleanest solution would be to do it just after the process has run. The problem is that if the workflow is lost in the case of remote use, the indexing is lost, even if the calculation has run well.

Normally not, hopefully: in "normal" operation mode, the server has a Soma-workflow database which can be queried to get the status of all pipelines and jobs, even afterwards when they are done or stopped. So in principle we can retrieve the status of jobs when we reconnect the MIA client to the computing server.
But it can sometimes happen that this database is erased/cleared/corrupted for some reasons (server crash, user removal of the workflows or of the complete database, bug, incompatible version change in soma-workflow, automatic cleanup after a timeout (1 week by default if I remember) etc), so we have to consider the possible situation.

**Comment by servoz** [comment]
Yes, but do we accept to live with an unmanaged abnormal situation ?
Why not, Microsoft has been doing it for decades with Windows ...! :-))
